File: p17/p1768/MergeAlternately.java
package leetcode.p17.p1768;

public class MergeAlternately {
    /**
     * Feb 11, 2024 00:34
     * Time Complexity: O(n+m)
     * Space Complexity: O(1)
     */
    public String mergeAlternately(String word1, String word2) {
        StringBuilder s = new StringBuilder();
        for (int i = 0; i < word1.length() && i < word2.length(); i++) {
            s.append(word1.charAt(i));
            s.append(word2.charAt(i));
        }
        if (s.length() / 2 == word1.length()) {
            s.append(word2.substring(s.length() / 2));
        } else {
            s.append(word1.substring(s.length() / 2));
        }
        return s.toString();
    }
}

File: p17/p1768/merge_alternately.py
class Solution:
    """
    Oct 19, 2024 23:41
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    def mergeAlternately(self, word1: str, word2: str) -> str:
        result = []

        for char1, char2 in zip(word1, word2):
            result.append(char1)
            result.append(char2)

        if len(word1) > len(word2):
            result.append(word1[len(word2):])
        else:
            result.append(word2[len(word1):])

        return "".join(result)


if __name__ == '__main__':
    s = Solution()
    s.mergeAlternately("1234", "abdfawfawf")
